pallets/nft-marketplace/src/lib.rs

```diff
@@ -0,0 +1,330 @@
+		type InstanceId: Parameter
+			+ Member
+			+ MaybeSerializeDeserialize
+			+ Copy
+			+ TypeInfo
+			+ From<u128>
+			+ IsType<InstanceIdOf<Self>>;
+
+		/// The Id of this pallet
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+	}
+
+	/// The gallery of NFTs currently for sale
+	#[pallet::storage]
+	#[pallet::getter(fn get_allowlisted)]
+	pub(super) type Gallery<T: Config> = StorageDoubleMap<
+		_,
+		// The hasher for the first key
+		Blake2_128Concat,
+		// The first key, the nft class Id
+		T::ClassId,
+		// The hasher for the second key
+		Blake2_128Concat,
+		// The second key, the nft instance id
+		T::InstanceId,
+		// The data regarding this item open for sale
+		SaleOf<T>,
+	>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// An NFT has been added to the gallery and is now for sale
+		ForSale(SaleOf<T>),
+
+		/// An NFT was removed from the gallery and is no longer for sale
+		Removed {
+			class_id: T::ClassId,
+			instance_id: T::InstanceId,
+		},
+
+		/// An NFT has been sold
+		Sold {
+			sale: SaleOf<T>,
+			buyer: T::AccountId,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// A user tried to add an NFT that could not be found
+		NotFound,
+
+		/// The origin is not the owner of an NFT
+		NotOwner,
+
+		/// A seller has attempted to list an NFT that is already for sale
+		AlreadyForSale,
+
+		/// An operation expected an NFT to be for sale when it is not
+		NotForSale,
+
+		/// A buyer attempted to buy an NFT they are selling
+		IsSeller,
+
+		/// The buyer does not have sufficient balance to buy the nft
+		InsufficientBalance,
+
+		/// Payment failed, i.e, failed to transfer the asking price from the buyer to the seller
+		PaymentFailed,
+
+		/// Failed to transfer an NFT from one account to another
+		FailedNftTransfer,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Add the given NFT to the gallery, putting it for sale.
+		///
+		/// Fails if
+		///   - the NFT is not found in [T::NonFungibles]
+		///   - `origin` is not the owner of the nft
+		///   - the nft is already for sale in the gallery
+		///   - transferring ownership of the NFT to this pallet's account fails
+		#[pallet::weight(10_000_000)]
+		pub fn add(
+			origin: OriginFor<T>,
+			class_id: T::ClassId,
+			instance_id: T::InstanceId,
+			currency: CurrencyOf<T>,
+			amount: BalanceOf<T>,
+		) -> DispatchResult {
+			let seller = ensure_signed(origin.clone())?;
+
+			// Check that the seller is the owner of the nft
+			ensure!(
+				Self::is_owner(seller.clone(), class_id, instance_id)?,
+				Error::<T>::NotOwner,
+			);
+
+			// Check that the nft is not yet for sale
+			ensure!(
+				!<Gallery<T>>::contains_key(class_id, instance_id),
+				Error::<T>::AlreadyForSale
+			);
+
+			// Transfer the NFT to the parachain account
+			T::NonFungibles::transfer(&class_id.into(), &instance_id.into(), &Self::account())
+				.map_err(|_| Error::<T>::FailedNftTransfer)?;
```

Comment:
As in other places, it's better to just return the error from the NFT pallet rather than hiding the details.